knebind/solver/solver.go

```diff
@@ -418,28 +528,72 @@ func (s *solver) genLinkRecurse(
 	first := keys[0]
 	for _, i := range m[first] {
 		if !used[i] {
+			grpResUpdated := false
+			// Ensure that for groups, all member links for a group get mapped to node links
+			// that are co-located in a single group
+			if first.group != "" {
+				if gName, ok := grpRes[first.group]; ok && gName != i.group {
+					// Some previous member of device port group got mapped to a different node intf group
+					continue
+				} else {
+					// This is to keep track of which iteration is adding entry to grpRes map,
+					// so as to delete it when res entry is deleted for that iteration
+					grpResUpdated = true
+				}
+			}
 			if !s.shouldLinkRecurse(d2n, first, i) {
 				continue
 			}
 			res[first] = i
+			grpRes[first.group] = i.group
 			used[i] = true
-			s.genLinkRecurse(d2n, m, keys[1:], res, used, ch)
+			s.genLinkRecurse(d2n, m, keys[1:], res, grpRes, used, ch)
 			delete(used, i)
+			delete(res, first)
+			if grpResUpdated {
+				// Remove only if this link was first in the group to add the map entry
+				delete(grpRes, first.group)
+			}
 		}
 	}
 }
 
-// shouldLinkRecurse returns whether a device link should be assigned to a node
-// link by checking the ports assigned match requirements.
+// arrangeLinkKeys creates a list of links (between a device pair) from testbed.
+// All regular links (with no group) are added at the end.
+// m - is the map of all testbed links and corresponding potential topology links for mapping
+// returns list of key links in the desirable processing order, with all regular links added at the end
+func (s *solver) arrangeLinkKeys(m map[*devLink][]*nodeLink) []*devLink {
```

Comment:
similarly, only include if a test or benchmark can distinguish between having and not having this logic

Reply:
Will add test for this too.